src/gui/processing/qgsprocessingwidgetwrapper.h

```diff
@@ -0,0 +1,212 @@
+     * The caller takes ownership of the returned widget. The wrapped widget can be retrieved at a later
+     * stage by calling wrappedWidget().
+     *
+     * The supplied \a context is used while initializing the widget to the parameter's default value.
+     *
+     * \see createWrappedLabel()
+     */
+    QWidget *createWrappedWidget( const QgsProcessingContext &context ) SIP_FACTORY;
+
+    /**
+     * Creates and returns a new label to accompany widgets created by the wrapper.
+     *
+     * The caller takes ownership of the returned label. Some parameter type and dialog type
+     * combinations will return nullptr for this method. If nullptr is returned, then no
+     * label should be shown for the parameter's widget (i.e. the label is embedded inside the
+     * widget itself).
+     *
+     * The wrapped label can be retrieved at a later stage by calling wrappedLabel().
+     *
+     * \see createWrappedWidget()
+     */
+    QLabel *createWrappedLabel() SIP_FACTORY;
+
+    /**
+     * Returns the current wrapped widget, if any.
+     * \see createWrappedWidget()
+     */
+    QWidget *wrappedWidget();
+
+    /**
+     * Returns the current wrapped label, if any.
+     * \see createWrappedLabel()
+     */
+    QLabel *wrappedLabel();
+
+    /**
+     * Returns the parameter definition associated with this wrapper.
+     */
+    const QgsProcessingParameterDefinition *parameterDefinition() const;
+
+    /**
+     * Sets the current \a value for the parameter.
+     *
+     * The \a context argument is used to specify the wider Processing context which the
+     * current value is associated with.
+     *
+     * \see value()
+     */
+    virtual void setWidgetValue( const QVariant &value, const QgsProcessingContext &context ) = 0;
+
+    /**
+     * Returns the current value of the parameter.
+     *
+     * \see setWidgetValue()
+     */
+    virtual QVariant value() const = 0;
+
+    /**
+     * Called after all wrappers have been created within a particular dialog or context,
+     * allowing the wrapper to connect to the wrappers of other, related parameters.
+     */
+    virtual void postInitialize( const QList< QgsAbstractProcessingParameterWidgetWrapper * > &wrappers );
+
+  protected:
+
+    /**
+     * Creates a new widget which allows customization of the parameter's value.
+     *
+     * The caller takes ownership of the returned widget.
+     *
+     * \see createLabel()
+     */
+    virtual QWidget *createWidget() = 0 SIP_FACTORY;
+
+    /**
+     * Creates a new label to accompany widgets created by the wrapper.
+     *
+     * The caller takes ownership of the returned label. Some parameter type and dialog type
+     * combinations will return nullptr for this method. If nullptr is returned, then no
+     * label should be shown for the parameter's widget (i.e. the label is embedded inside the
+     * widget itself).
+     *
+     * \see createWidget()
+     */
+    virtual QLabel *createLabel() SIP_FACTORY;
```

Comment:
I'm wondering whether it'd be worth adding a virtual QIcon icon() function here, in case we'd ever want to provide icon representation of parameters (which would be useful for the model). For now, we could have a default icon, which would be use in the modeler (ie we can take the + icon already used).

Reply:
Wouldn't that be in the parameter definition, not the widget definition?